### PR TITLE
bugfix/24039-bubble-marker

### LINKED ISSUE #24039: Bubble with a maximum size percentage are not completely drawn inside the plot area.
#### Expected behaviour
I expect that the axis scale will be adjusted such that all the bubbles can be completely contained within the plot area

#### Actual behaviour
The bubble to the right is overlapping the x axis.

Image

#### Live demo with steps to reproduce
https://codepen.io/ZenRanger1979/pen/NPrpGKV

### PR DESCRIPTION
**Pull request by magdalena-gut**
Fixed #24039, the bubble with a maximum size percentage was clipped.

---
- To see the specific tasks where the Asana app for GitHub is being used, see below:
  - https://app.asana.com/0/0/1212842649102701

### CODE REVIEW

**Comment by magdalena-gut** [comment]
_Internal note:_ 
I'm not sure about this visual regression - whether the differences are significant enough to accept or not. I'd appreciate some feedback.

**Review by hubertkozik — CHANGES_REQUESTED** [review]
Visual test bot found a diff in `highcharts/bubble-legend/connector` (plain default-settings sample, not an edge case) and you flagged uncertainty about it yourself with no follow-up since — that needs resolving before merge.

Verified the fix itself works: built both branches and ran the new test headlessly — it genuinely fails on master (~40px overflow) and passes cleanly on this branch, so this isn't a false-positive-passes-on-master test.

A few implementation notes left as inline comments.

File: ts/Series/Bubble/BubbleSeries.ts [hunk 1]
```diff
@@ -177,8 +173,101 @@ function onAxisFoundExtremes(
...
+    }
+
+    // Only correct if tick snap shifted the axis after onAxisFoundExtremes
+    // already padded it — otherwise we cause a double expansion. #24039
+    if (
+        this.min === (this as any).bubbleMinAfterPadding &&
+        this.max === (this as any).bubbleMaxAfterPadding
+    ) {
+        return;
     }
```
Comment (hubertkozik) [inline ts/Series/Bubble/BubbleSeries.ts]: This guard compares `this.min`/`this.max` against values stashed in `foundExtremes`, but softMin/softMax, single-point expansion, and `adjustTickAmount` can also change extrema before `afterSetTickPositions` fires, so this detects "anything changed since foundExtremes", not tick-snap specifically. Harmless since the correction is a no-op when bubbles already fit, but worth tightening or re-commenting to match what it actually checks.

File: ts/Series/Bubble/BubbleSeries.ts [hunk 2]
```diff
@@ -177,8 +173,101 @@ function onAxisFoundExtremes(
...
+
+        const data = series.getColumn(isXAxis ? 'x' : 'y');
+
+        let i = data.length;
+        while (i--) {
+            const d = data[i];
+            if (!isNumber(d)) {
+                continue;
+            }
+            const r = series.radii && series.radii[i] || 0;
```
Comment (hubertkozik) [inline ts/Series/Bubble/BubbleSeries.ts]: `series.radii` is read directly here, while the sibling `onAxisFoundExtremes` resolves it through `(series.onPoint || series).getRadii()` first (bubbles rendered via `onPoint`, e.g. tilemap overlays). This pass implicitly depends on the x-axis pass having already refreshed `series.radii`.

**Review by hubertkozik — COMMENTED** [review]
Inline follow-ups to my previous review.

File: ts/Series/Bubble/BubbleSeries.ts [hunk 3]
```diff
@@ -177,8 +173,101 @@ function onAxisFoundExtremes(
                 (this as any)[keys[0]] += keys[2] / transA;
             }
         });
+
+    }
+
+    // Store values so afterSetTickPositions can detect tick snap. #24039
+    if (hasActiveSeries) {
+        (this as any).bubbleMinAfterPadding = this.min;
+        (this as any).bubbleMaxAfterPadding = this.max;
```
Comment (hubertkozik) [inline ts/Series/Bubble/BubbleSeries.ts]: `bubbleMinAfterPadding`/`bubbleMaxAfterPadding` go through `(this as any)` instead of the module-augmentation pattern already used in this file for `bubblePadding`/`radii`.

File: ts/Series/Bubble/BubbleSeries.ts [hunk 4]
```diff
@@ -177,8 +185,111 @@ function onAxisFoundExtremes(
...
+            }
+            const range = axisMax - axisMin,
+                pxPos = range > 0 ?
+                    (d - axisMin) / range * axisLength :
+                    axisLength / 2;
+
+            if (!hasUserMin && pxPos - r < -1) {
+                requiredMin = Math.min(
+                    requiredMin,
+                    (r * axisMax - d * axisLength) / (r - axisLength)
```
Comment (hubertkozik) [inline ts/Series/Bubble/BubbleSeries.ts]: `requiredMin`/`requiredMax` are each solved against the pre-correction `axisMin`/`axisMax` but applied together — when bubbles overflow both ends at once this under-corrects (two points with `r` = 30% of `axisLength` at opposite ends leave ~7px residual overflow after both `this.min`/`this.max` are set, per this file's own formulas). Needs to solve jointly or re-validate after applying both.

File: ts/Series/Bubble/BubbleSeries.ts [hunk 5]
```diff
@@ -177,8 +185,111 @@ function onAxisFoundExtremes(
...
+            radii = (series.onPoint || series).radii;
+
+        let i = data.length;
+        while (i--) {
+            const d = data[i];
+            if (!isNumber(d)) {
+                continue;
+            }
+            const r = radii && radii[i] || 0;
+            if (r >= axisLength) {
```
Comment (hubertkozik) [inline ts/Series/Bubble/BubbleSeries.ts]: Only guards exact `r >= axisLength`; for `r` just under `axisLength` (large `maxSize` in a narrow plot area) `axisLength - r` in the `requiredMax` calc below approaches 0 and blows up the extreme — worth widening the guard and adding a test for that case.

File: ts/Series/Bubble/BubbleSeries.ts [hunk 6]
```diff
@@ -177,8 +185,111 @@ function onAxisFoundExtremes(
...
+            }
+        }
+    });
+
+    if (!hasActiveSeries) {
+        return;
     }
 
+    if (!hasUserMin && requiredMin < axisMin) {
+        this.min = requiredMin;
```
Comment (hubertkozik) [inline ts/Series/Bubble/BubbleSeries.ts]: `setAxisTranslation()` already ran for this pass before `afterSetTickPositions` fires, so mutating `this.min`/`this.max` here doesn't refresh `transA`/`translationSlope` used by `Axis.translate()` in the same pass — is the correction relying on a later `setScale()` pass to actually take effect? Might explain the still-open `bubble-legend/connector` diff.

**Review by hubertkozik — CHANGES_REQUESTED** [review]
Visual test bot still flags a diff in `highcharts/bubble-legend/connector` (14,278 px, unrelated to this PR's own test changes) — you raised uncertainty about this yourself last month with no reply since; needs resolving before merge.

Verified `samples/unit-tests/series-bubble/minmaxsize/demo.js` genuinely fails on master and passes here, so the regression test itself is solid.

Found a correctness gap in the new double-overflow handling in `onAxisAfterSetTickPositions` — left as inline comments.

**Comment by magdalena-gut** [comment]
Ended up reworking this instead of fixing those one by one — they were all the same problem underneath, so that code is gone now.

Regarding the visual diff, it actually looks like the fix working, as master really does clip a bubble by ~4px there. The 14,278px is just xAxis.min moving 0.57 units and shifting everything with it.

On `transA` — it does get recalculated via `setChartSize`, so that's fine. `axis.len` is the stale one, which is why this gets most of the clipping, not all of it.

File: ts/Series/Bubble/BubbleSeries.ts [hunk 7]
```diff
@@ -180,6 +181,115 @@ function onAxisFoundExtremes(
...
+            tickPositions.length > 0 &&
+            !this.tickAmount &&
+            !this.categories &&
+            !this.dateTime
+        ),
+        snapStart = canSnap && options.startOnTick,
+        snapEnd = canSnap && options.endOnTick,
+        tickSteps = (value: number, atEnd: boolean): number => Math.max(
+            Math.ceil(
+                (atEnd ? value - lastTick : firstTick - value) / tickInterval
```
Comment (hubertkozik) [inline ts/Series/Bubble/BubbleSeries.ts]: Nit: this division isn't epsilon-guarded before `Math.ceil` — `(0.1 - -0.2) / 0.1` evaluates to `3.0000000000000004` in JS, so `Math.ceil` returns `4` instead of `3` and adds one unneeded extra tick step for some decimal `tickInterval`s.

File: ts/Series/Bubble/BubbleSeries.ts [hunk 8]
```diff
@@ -180,6 +181,115 @@ function onAxisFoundExtremes(
...
+
+    if (!isNumber(paddedMin) || !isNumber(paddedMax)) {
+        return;
+    }
+
+    // Growing by whole ticks keeps `startOnTick` and `endOnTick` true, but
+    // only where they are evenly spaced and no tick count has been settled on
+    const firstTick = tickPositions[0],
+        lastTick = tickPositions[tickPositions.length - 1],
+        canSnap = (
```
Comment (hubertkozik) [inline ts/Series/Bubble/BubbleSeries.ts:205]: Nit: `canSnap` doesn't exclude `options.tickPositions`/`tickPositioner` — `tickInterval` is still computed even when an explicit tick list or custom positioner overrides the grid, so a bubble axis combined with either could synthesize ticks unrelated to the actual custom positions.

File: ts/Series/Bubble/BubbleSeries.ts [hunk 9]
```diff
@@ -529,6 +639,11 @@ class BubbleSeries extends ScatterSeries {
 
         if (pushUnique(composed, 'Series.Bubble')) {
             addEvent(AxisClass, 'foundExtremes', onAxisFoundExtremes);
+            addEvent(
+                AxisClass,
+                'afterSetTickPositions',
```
Comment (hubertkozik) [inline ts/Series/Bubble/BubbleSeries.ts:655]: Nit: `BrokenAxis` also listens on `afterSetTickPositions` and filters ticks inside breaks — worth double-checking this runs after that listener, since otherwise the ticks pushed/unshifted below could land inside a break.

File: ts/Series/Bubble/BubbleSeries.ts [hunk 10]
```diff
@@ -180,6 +181,115 @@ function onAxisFoundExtremes(
...
+        snapStart = canSnap && options.startOnTick,
+        snapEnd = canSnap && options.endOnTick,
+        tickSteps = (value: number, atEnd: boolean): number => Math.max(
+            Math.ceil(
+                (atEnd ? value - lastTick : firstTick - value) / tickInterval
+            ),
+            0
+        );
+
+    let passes = 5;
```
Comment (hubertkozik) [inline ts/Series/Bubble/BubbleSeries.ts:232]: Nit: for a bubble radius near ~50% of the axis length, 5 passes may not fully converge before the 2x give-up threshold trips — might be worth a test in that range.

**Review by hubertkozik — APPROVED** [review]
Rework fixes the double-overflow math bug from the previous round, and the `bubble-legend/connector` visual diff checks out — built both branches and measured it, master clips ~4px there, this branch is within 0.01px, so that's the fix working, not a regression.

A few nits below, nothing blocking.

File: ts/Series/Bubble/BubbleSeries.ts (hunk 10 above)
Comment (magdalena-gut) [reply ts/Series/Bubble/BubbleSeries.ts:232]: I tried more passes - it helps those bubbles, but breaks the maxSize: '100%' case, so kept 5
